src/main/java/frc/robot/poseestimator/VisionDenoiser.java

```diff
@@ -0,0 +1,142 @@
+package frc.robot.poseestimator;
+
+import edu.wpi.first.math.filter.LinearFilter;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.interpolation.TimeInterpolatableBuffer;
+import frc.robot.poseestimator.observations.VisionObservation;
+import frc.utils.time.TimeUtils;
+
+import java.util.*;
+
+public class VisionDenoiser {
+
+	private final int maximumSize;
+	private LinkedList<VisionObservation> observations = new LinkedList<>();
+	private LinearFilter xFilter;
+	private LinearFilter yFilter;
+	private LinearFilter angleFilterRadians;
+	private Pose2d lastFilterOutput;
+
+	public VisionDenoiser(int maximumSize) {
+		this.maximumSize = maximumSize;
+
+		this.xFilter = LinearFilter.movingAverage(maximumSize);
+		this.yFilter = LinearFilter.movingAverage(maximumSize);
+		this.angleFilterRadians = LinearFilter.movingAverage(maximumSize);
+	}
+
+	private void popLastIfQueueTooLarge() {
+		if (observations.size() > maximumSize) {
+			observations.poll();
+		}
+	}
+
+	public void addVisionObservation(VisionObservation observation) {
+		observations.add(observation);
+		popLastIfQueueTooLarge();
+		lastFilterOutput = getFilterResult(
+			xFilter.calculate(observation.robotPose().getX()),
+			yFilter.calculate(observation.robotPose().getY()),
+			angleFilterRadians.calculate(observation.robotPose().getRotation().getRadians())
+		);
+	}
+
+	private ArrayList<Pose2d> getRobotPosesFromGivenObservations() {
+		ArrayList<Pose2d> output = new ArrayList<>();
+		for (VisionObservation observation : observations) {
+			output.add(observation.robotPose());
+		}
+		return output;
+	}
+
+	private Optional<VisionObservation> calculateAverage(LinkedList<VisionObservation> overwrittenObservations) {
+		if (!overwrittenObservations.isEmpty()) {
+			return Optional.empty();
+		}
+		ArrayList<Pose2d> poses = getRobotPosesFromGivenObservations();
+		return Optional.of(
+			new VisionObservation(
+				PoseEstimationMath.meanOfPose(poses),
+				PoseEstimationMath.calculateStandardDeviationOfPose(poses),
+				overwrittenObservations.peekLast().timestamp()
+			)
+		);
+	}
+
+	public Optional<VisionObservation> calculateAverage() {
+		return calculateAverage(observations);
+	}
+
+	public Optional<VisionObservation>
+		calculateAverageFixedPose(Pose2d currentPose, Pose2d odometryPose, TimeInterpolatableBuffer<Pose2d> odometryPoseInterpolator) {
+		return calculateAverage(fixAccordingToOdometry(currentPose, odometryPose, odometryPoseInterpolator));
+	}
+
+	private Optional<VisionObservation> calculateWeightedAverage(LinkedList<VisionObservation> overwrittenObservations) {
```

Comment:
```suggestion
	private Optional<VisionObservation> calculateWeightedAverage(List<VisionObservation> overwrittenObservations) {
```

Reply:
I need support of the peekLast method (reduces runtime too) so it's better to be that way

Reply:
so, Queue?

Reply:
Queue is an interface of LinkedList

Reply:
Queue doesn't support that too. Though I do need to learn Java's data structures after the off-season.